### PR TITLE
38312: Aggregations page

### PR DESCRIPTION
**Pull request by mcmorisi**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-38312>
Staging - <https://preview-mongodbmcmorisi.gatsbyjs.io/csharp/38312-aggregations/fundamentals/aggregation/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

**Comment by rishitb-mongodb** [comment]
Some suggestions: 
- There will rarely be any users using BSON for simple aggregation operations. Can we show users how to perform the same aggregation [example](https://preview-mongodbmcmorisi.gatsbyjs.io/csharp/38312-aggregations/fundamentals/aggregation/#aggregation-example) using Builders and LINQ below that?
- In the Additional Information section, just like LINQ, can we reference aggregation using [Builders](https://preview-mongodbmcmorisi.gatsbyjs.io/csharp/38312-aggregations/fundamentals/builders/#build-an-aggregation-pipeline) too?

File: source/fundamentals/aggregation.txt [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to use the {+driver-long+} to perform
+**aggregation operations**.
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:24]: S: redo this section to align more with the page structure on the [Rust aggregation page](https://github.com/mongodb/docs-rust/blob/master/source/fundamentals/aggregation.txt)

That page has a better organization of information and the copy has been updated to align more with the style guide

File: source/fundamentals/aggregation.txt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  of 16 megabytes.
+
+- Pipeline stages have a memory limit of 100 megabytes by default. If required, you can exceed this limit by setting
+  the `AllowDiskUse <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.AggregateOptions.AllowDiskUse.html#MongoDB_Driver_AggregateOptions_AllowDiskUse>`__
+  property of the ``AggregateOptions`` object that you pass to the ``Aggregate()`` method.
+
+  .. important:: ``$graphLookup`` exception
+
+     The :manual:`$graphLookup </reference/operator/aggregation/graphLookup/>` stage has
+     a strict memory limit of 100 megabytes and will ignore the ``AllowDiskUse`` property.
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:71]:
```suggestion
     The :manual:`$graphLookup </reference/operator/aggregation/graphLookup/>` stage has
     a strict memory limit of 100 megabytes and ignores the ``AllowDiskUse`` property.
```

File: source/includes/fundamentals/code-examples/Aggregation.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    public static void Main(string[] args)
+    {   
+        var mongoClient = new MongoClient(MongoConnectionString);
+        
+        var database = mongoClient.GetDatabase("sample_restaurants");
+        var collection = database.GetCollection<BsonDocument>("restaurants");
+
+        // begin-aggregation
+        // Defines our $match and $group aggregation stages
```
Comment (rustagir) [inline source/includes/fundamentals/code-examples/Aggregation.cs:19]:
```suggestion
        // Defines the $match and $group aggregation stages
```

**Review by rustagir — CHANGES_REQUESTED** [review]
a few things!

File: source/fundamentals/aggregation.txt [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  a strict memory limit of 100 megabytes and ignores the ``AllowDiskUse`` property.
+
+Aggregation Example
+-------------------
+
+.. note::
+  
+  This example uses the ``sample_restaurants.restaurants`` collection
+  from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
+  free MongoDB Atlas cluster and load the sample datasets, see :ref:`csharp-quickstart`.
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:99]: S: move this note below the following paragraph

File: source/fundamentals/aggregation.txt [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+- A :manual:`$match </reference/operator/aggregation/match/>` stage to filter for documents whose
+  ``cuisine`` field contains the value ``"Bakery"``.
+
+- A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching
+  documents by the ``borough`` field, accumulating a count of documents for each distinct value
+  of that field.
+
+We have provided three versions of this example that each provide different approaches to
+defining the aggregation stages used in the example pipeline. The first uses 
+``BsonDocument`` objects, the second uses LINQ, and the third uses builders.
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:116]:
```suggestion
The following tabs contain versions of this example that each provide different approaches to
create and combine the aggregation stages used in the example pipeline. Select from the following :guilabel:`BsonDocument`, :guilabel:`LINQ`, and :guilabel:`Builders` tabs to view an example that uses each API to perform the aggregation.
```

File: source/fundamentals/aggregation.txt [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+.. tabs::
+
+   .. tab:: ``BsonDocument`` approach
+      :tabid: bsondocument
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:121]: S: remove monospacing from header elements, Also I think that this tab should be last given the technical feedback
```suggestion
   .. tab:: BsonDocument
      :tabid: bsondocument approach
```

Comment (rustagir) [reply source/fundamentals/aggregation.txt:121]: same formatting suggestion applies to all three examples

File: source/fundamentals/aggregation.txt [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+            { _id = Bronx, Count = 71 }
+            { _id = Brooklyn, Count = 173 }
+            { _id = Staten Island, Count = 20 }
+            { _id = Missing, Count = 2 }
+            { _id = Manhattan, Count = 221 }
+            { _id = Queens, Count = 204 }      
+
+To learn more about using LINQ and Builders to construct aggregation pipelines,
+see :ref:`csharp-linq` and :ref:`Builders <csharp-builders-aggregation>`.
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:187]:
```suggestion
To learn more about using LINQ and Builders to construct aggregation pipelines,
see the :ref:`csharp-linq`guide and the :ref:`csharp-builders-aggregation` section of the Operations with Builders guide.
```

File: source/fundamentals/aggregation.txt [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+- A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching
+  documents by the ``borough`` field, accumulating a count of documents for each distinct value
+  of that field.
+
+We have provided three versions of this example that each provide different approaches to
+defining the aggregation stages used in the example pipeline. The first uses 
+``BsonDocument`` objects, the second uses LINQ, and the third uses builders.
+
+.. tabs::
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:118]: S: would having subsections for these sections be a better alternative to tabs? Sometimes, we avoid putting code examples in tabs as it reduces the searchability of the nested content.

**Review by rustagir — CHANGES_REQUESTED** [review]
almost there! a few more items

File: source/fundamentals/aggregation.txt (hunk 5 above)
Comment (mcmorisi) [reply source/fundamentals/aggregation.txt:116]: Will take a pass at rewording this while also moving away from the tabs approach.

File: source/fundamentals/aggregation.txt (hunk 8 above)
Comment (mcmorisi) [reply source/fundamentals/aggregation.txt:118]: Will split into subsections

File: source/fundamentals/aggregation.txt [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+       | Count the results
+     - | Select *certain* documents to return
+       | Select *which* fields to return
+       | Sort the results
+       | Limit the results
+       | Count the results
+       | Rename fields
+       | Compute new fields
+       | Summarize data
+       | Connect and merge data sets
```
Comment (BorisDog) [inline source/fundamentals/aggregation.txt:74]: Add "Group the results".

File: source/fundamentals/aggregation.txt [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+Consider the following :manual:`limitations </core/aggregation-pipeline-limits/>` when 
+performing aggregation operations:
+
+- Returned documents must not violate the :manual:`BSON document size limit </reference/limits/#mongodb-limit-BSON-Document-Size>`
+  of 16 megabytes.
+
+- Pipeline stages have a memory limit of 100 megabytes by default. If required, you can exceed this limit by setting
+  the `AllowDiskUse <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.AggregateOptions.AllowDiskUse.html#MongoDB_Driver_AggregateOptions_AllowDiskUse>`__
+  property of the ``AggregateOptions`` object that you pass to the ``Aggregate()`` method.
```
Comment (BorisDog) [inline source/fundamentals/aggregation.txt:87]: Do we want link to API docs here for `AggregateOptions` and possibly `Aggregate`?

File: source/includes/fundamentals/code-examples/LinqAggregation.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+        var database = mongoClient.GetDatabase("sample_restaurants");
+        var collection = database.GetCollection<Restaurant>("restaurants");
+
+        // begin-aggregation
+        // Defines a queryable collection object as a prerequisite to using LINQ
+        var queryableCollection = collection.AsQueryable();
+
+        // Executes the $match and $group aggregation stages
+        var query = queryableCollection
```
Comment (BorisDog) [inline source/includes/fundamentals/code-examples/LinqAggregation.cs:24]: styling: Indentation should be  (4 spaces)
```
var query = queryableCollection
    .Where(...)...
```

File: source/includes/fundamentals/code-examples/LinqAggregation.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        var mongoClient = new MongoClient(MongoConnectionString);
+
+        var database = mongoClient.GetDatabase("sample_restaurants");
+        var collection = database.GetCollection<Restaurant>("restaurants");
+
+        // begin-aggregation
+        // Defines a queryable collection object as a prerequisite to using LINQ
+        var queryableCollection = collection.AsQueryable();
+
+        // Executes the $match and $group aggregation stages
```
Comment (BorisDog) [inline source/includes/fundamentals/code-examples/LinqAggregation.cs:23]: Just to be precise, this statement defines the query, and the next `ToList` call actually executes it.

File: source/includes/fundamentals/code-examples/Aggregation.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        
+        var database = mongoClient.GetDatabase("sample_restaurants");
+        var collection = database.GetCollection<BsonDocument>("restaurants");
+
+        // begin-aggregation
+        // Defines the $match and $group aggregation stages
+        var matchStage = new BsonDocument
+        {
+            {
+                "$match", new BsonDocument {
```
Comment (BorisDog) [inline source/includes/fundamentals/code-examples/Aggregation.cs:23]: Alman style brackets:

```
var matchStage = new BsonDocument
{
    {
        "$match", 
        new BsonDocument 
        {
            { "cuisine", "Bakery" }
        }
    }
};

```

File: source/includes/fundamentals/code-examples/Aggregation.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                "$group", new BsonDocument {
+                    { "_id", "$borough" },
+                    { "count", new BsonDocument("$sum", 1) }
+                }
+            }
+        };
+
+        // Executes the aggregation pipeline
+        var pipeline = new BsonDocument[] { matchStage, groupStage };
```
Comment (BorisDog) [inline source/includes/fundamentals/code-examples/Aggregation.cs:40]: nit: `var pipeline = new[] { matchStage, groupStage };`

File: source/includes/fundamentals/code-examples/Aggregation.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        var results = collection.Aggregate<BsonDocument>(pipeline).ToList();
+
+        // Prints the aggregated results
+        foreach(BsonDocument result in results)
```
Comment (BorisDog) [inline source/includes/fundamentals/code-examples/Aggregation.cs:44]: `foreach (var result)` (add space after foreach)

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: source/fundamentals/aggregation.txt (hunk 10 above)
Comment (mcmorisi) [reply source/fundamentals/aggregation.txt:87]: We have them linked below in the API Documentation subsection of the Additional Information section.

File: source/fundamentals/aggregation.txt [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+``IMongoCollection<TDocument>.Aggregate()`` method.
+
+.. note::
+  
+  This example uses the ``sample_restaurants.restaurants`` collection
+  from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
+  free MongoDB Atlas cluster and load the sample datasets, see :ref:`csharp-quickstart`.
+
+The following code example produces a count of the number of bakeries in each borough
+of New York. To do so, it uses an aggregation pipeline with the following stages:
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:106]:
```suggestion
The following code example produces a count of the number of bakeries in each borough
of New York City. To do so, it uses an aggregation pipeline that contains the following stages:
```

File: source/fundamentals/aggregation.txt [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+- A :manual:`$match </reference/operator/aggregation/match/>` stage to filter for documents whose
+  ``cuisine`` field contains the value ``"Bakery"``.
+
+- A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching
+  documents by the ``borough`` field, accumulating a count of documents for each distinct value
+  of that field.
+
+The following sections contain versions of this example that each provide different 
+approaches to creating and combining the aggregation stages used in the example pipeline.
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:116]:
```suggestion
The following sections implement this example by using LINQ, Builders, and BsonDocument 
approaches to create and combine the aggregation stages used in the example pipeline.
```

File: source/fundamentals/aggregation.txt [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+      :language: console
+      :visible: false
+
+      { "_id" : "Brooklyn", "count" : 173 }
+      { "_id" : "Manhattan", "count" : 221 }
+      { "_id" : "Bronx", "count" : 71 }
+      { "_id" : "Missing", "count" : 2 }
+      { "_id" : "Staten Island", "count" : 20 }
+      { "_id" : "Queens", "count" : 204 }
+
```
Comment (rustagir) [inline source/fundamentals/aggregation.txt:189]: S: add a note here explaining why users should prefer to use one of the other APIs

**Review by rustagir — APPROVED** [review]
approving with some suggestions. Happy to take another look if needed!

**Review by BorisDog — APPROVED** [review]
LGTM!

**Pull request merged**